### PR TITLE
spice.ai table provider

### PR DESCRIPTION
**Pull request by y-f-u**
NOTE: this still doesn't support `count` and `distinct` etc. Which will work on after this one merged. Given it will be the same logic to enable all table providers we have now

With this change, spice.ai and dremio data source can be queried without acceleration. The query will go to the remote and fetched via flight.

e.g.  with this yaml
```yaml
kind: Spicepod
name: demo
version: v1beta1

datasets:
- from: spice.ai/eth.recent_blocks
  name: eth_blocks
  # acceleration:
  #   enabled: true
  #   refresh_mode: full
- from: spice.ai/eth.recent_blocks
  name: blocks
  acceleration:
    enabled: true
    refresh_mode: full
- name: test_view
  sql: select number as ts, (number::double / 100) as y, (number::double) / 100 as y2 from eth_blocks order by ts desc limit 100
```

queries like this can be done:
```bash
sql> select number from eth_blocks where number not in (select number from blocks);
+----------+
| number   |
+----------+
| 19281206 |
| 19281205 |
| 19281204 |
| 19281203 |
| 19281201 |
| 19281202 |
| 19281199 |
+----------+

Query took: 0.971788792 seconds
```

### CODE REVIEW

File: crates/flight_datafusion/src/lib.rs
Comment (mitchdevenport) [inline crates/flight_datafusion/src/lib.rs]: Don't clone the client directly, use an Arc to share it between threads

File: crates/flight_datafusion/src/lib.rs
Comment (y-f-u) [inline crates/flight_datafusion/src/lib.rs:1]: This file still has duplication with sql_provider_datafusion.

The idea is to wait until MItch's branch merged, then work on another refactoring. As, for flight client, there is no need for the concept the ConnectionPool, also it doesn't need to have the implement DbManager in SQLTable too.

Merging and refactoring now may cause stepping over other's toes. So please ignore the duplication.

File: crates/flight_datafusion/src/lib.rs
Comment (y-f-u) [inline crates/flight_datafusion/src/lib.rs]: arc doesn't work well with async

Comment (y-f-u) [reply crates/flight_datafusion/src/lib.rs]: intentional copy

File: crates/runtime/src/dataconnector/debug.rs
Comment (phillipleblanc) [inline crates/runtime/src/dataconnector/debug.rs]: no need to implement this if the default implementation is the same

File: crates/runtime/src/datafusion.rs
Comment (phillipleblanc) [inline crates/runtime/src/datafusion.rs]:
```suggestion
    pub fn attach_mesh(
```

File: crates/flight_datafusion/src/lib.rs
Comment (phillipleblanc) [inline crates/flight_datafusion/src/lib.rs]: There is no point wrapping a FlightClient in an Arc - because as you've discovered, we need to clone it to be able to query. I would just remove the Arc<>

Comment (phillipleblanc) [reply crates/flight_datafusion/src/lib.rs]: The real way to fix this would be using a connection pooler - should be fairly straightforward using r2d2 - but let's track that separately

File: crates/flight_datafusion/src/lib.rs
Comment (phillipleblanc) [inline crates/flight_datafusion/src/lib.rs]:
```suggestion
    client: FlightClient,
```

File: crates/flight_datafusion/src/lib.rs
Comment (phillipleblanc) [inline crates/flight_datafusion/src/lib.rs]: The naming here is weird - the sources you are connecting to aren't FlightSQL endpoints - they are normal Flight endpoints that support SQL queries using the `Cmd` FlightDescriptors - which are two separate things.

Mitch is implementing a real `FlightSQL` provider - so I think we should rename this to `FlightTable`

File: crates/flight_datafusion/src/lib.rs
Comment (phillipleblanc) [reply crates/flight_datafusion/src/lib.rs:1]: As mentioned in another comment - we will eventually need the connection pool for FlightClient

File: crates/flight_datafusion/src/lib.rs
Comment (phillipleblanc) [inline crates/flight_datafusion/src/lib.rs]:
```suggestion
        write!(f, "FlightExec sql={sql}")
```

File: crates/flight_datafusion/src/lib.rs
Comment (phillipleblanc) [inline crates/flight_datafusion/src/lib.rs]:
```suggestion
        write!(f, "FlightExec sql={sql}")
```

File: crates/flight_datafusion/src/lib.rs
Comment (phillipleblanc) [inline crates/flight_datafusion/src/lib.rs]: We shouldn't do this - make `get_schema` and `new` async

Comment (phillipleblanc) [reply crates/flight_datafusion/src/lib.rs]: this means you will also need to make `get_table_provider` and `attach_mesh` async as well.

File: crates/flight_datafusion/src/lib.rs
Comment (y-f-u) [reply crates/flight_datafusion/src/lib.rs]: sounds good, can be done when refactoring them.

File: crates/flight_datafusion/src/lib.rs
Comment (y-f-u) [reply crates/flight_datafusion/src/lib.rs:1]: yeah will do in the refactoring

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**
